src/packet/tcp.rs

```diff
@@ -0,0 +1,197 @@
+       TCP headers must have data offset set to 5 or more.
+     */
+    let mut v = tcp.get_data_offset_and_reserved();
+    v  = v >> 4; // get rid of reserved bits
+    v = v - 5; // remove the size of the minimal tcp header of 20 bytes
+    v = v * 4; // multiple the remaining word count by 4 to get the byte count
+    return v as usize;
+}
+
+/// Represents the TCP Option fields
+#[packet]
+pub struct TcpOptions {
+    kind: u8,
+    #[payload]
+    data: Vec<u8>
+}
+
+fn tcp_padding_length<'a>(tcp: &TcpPacket<'a>) -> usize {
+    /* The TCP header padding is used to ensure that the entire header
+       ends on a 32 bit boundary.
+     */
+    //return tcp.get_data_offset_and_reserved() as usize % 4;
+    return 0
+}
+
+/// Represents the TCP header padding
+#[packet]
+pub struct TcpPadding {
+    #[payload]
+    data: Vec<u8>
+}
+
+/// Calculates the checksum of a TCP packet
+/// The passed in TcpPacket must have it's initial checksum value set to zero.
+pub fn checksum<'a, T: HasPseudoheader>(packet: &TcpPacket<'a>, encapsulating_packet: T) -> u16be {
+    let mut sum = encapsulating_packet.pseudoheader_checksum();
+    let length = packet.packet().len() as u32;
+    sum = sum + length & 0xffff;
+    sum = sum + length >> 16;
+    return rfc1071_checksum(packet.packet(), sum);
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const TCP_MIN_HEADER_LEN: usize = 20;
+    const TCP_OPTIONS_LEN: usize = 12;
+    const PAYLOAD_LEN: usize = 5;
+
+    #[test]
+    fn tcp_header_ipv4_test() {
+        use pnet::packet::ip::IpNextHeaderProtocols;
+        use pnet::packet::ipv4::{Ipv4Packet,MutableIpv4Packet};
+        use std::net::{Ipv4Addr};
+
+        const IPV4_HEADER_LEN: usize = 20;
+
+        let mut packet = [0u8; IPV4_HEADER_LEN + TCP_MIN_HEADER_LEN + TCP_OPTIONS_LEN + PAYLOAD_LEN];
+        let ipv4_source = Ipv4Addr::new(127, 0, 0, 1);
+        let ipv4_destination = Ipv4Addr::new(127, 0, 0, 1);
+        let next_level_protocol = IpNextHeaderProtocols::Tcp;
+        let mut csum = 0;
+
+        {
+            let mut mut_ip_header = MutableIpv4Packet::new(&mut packet[..]).unwrap();
+            mut_ip_header.set_next_level_protocol(next_level_protocol);
+            mut_ip_header.set_source(ipv4_source);
+            mut_ip_header.set_destination(ipv4_destination);
+        }
+
+        {
+            let mut mutable_tcp_header = MutableTcpPacket::new(&mut packet[IPV4_HEADER_LEN..]).unwrap();
+            mutable_tcp_header.set_checksum(0);
+        }
+
+        {
+            let ip_header = Ipv4Packet::new(&packet[..IPV4_HEADER_LEN]).unwrap();
+            let tcp_header = TcpPacket::new(&packet[IPV4_HEADER_LEN..]).unwrap();
+            csum = checksum(&tcp_header, ip_header);
+        }
+
+        {
+            let mut mutable_tcp_header = MutableTcpPacket::new(&mut packet[IPV4_HEADER_LEN..]).unwrap();
+            mutable_tcp_header.set_checksum(csum);
```

Comment:
It's very unfortunate that setting a checksum is this verbose - it would be great if this could be improved.